core/src/test/java/kafka/server/ProducerRebootstrapIntegrationTest.java

```diff
@@ -0,0 +1,267 @@
+import org.apache.kafka.common.test.api.ClusterTestDefaults;
+import org.apache.kafka.common.test.api.ClusterTestExtensions;
+import org.apache.kafka.common.test.api.Type;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.coordinator.group.GroupCoordinatorConfig;
+import org.apache.kafka.server.config.ReplicationConfigs;
+
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@Timeout(240)
+@ExtendWith(ClusterTestExtensions.class)
+@ClusterTestDefaults(
+        types = {Type.KRAFT},
+        brokers = 2,
+        serverProperties = {
+            @ClusterConfigProperty(key = TopicConfig.UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG, value = "true"),
+            @ClusterConfigProperty(key = ReplicationConfigs.UNCLEAN_LEADER_ELECTION_INTERVAL_MS_CONFIG, value = "1"),
+        }
+
+)
+public class ProducerRebootstrapIntegrationTest {
+    final String topicName = "topic";
+    final Map<Integer, List<Integer>> expectedReplicaAssignment = Map.of(0, List.of(0, 1), 1, List.of(0, 1));
+    private static final int[] PORTS = choosePorts(2);
+    
+    static {
+        // Set system properties that will be used to replace ${port0} placeholders
+        System.setProperty("port0", String.valueOf(PORTS[0]));
+        System.setProperty("port1", String.valueOf(PORTS[1]));
+    }
+
+    @Timeout(240)
+    @ClusterTest(
+        types = {Type.KRAFT},
+        brokers = 2,
+        serverProperties = {
+            @ClusterConfigProperty(key = TopicConfig.UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG, value = "true"),
+            @ClusterConfigProperty(key = ReplicationConfigs.UNCLEAN_LEADER_ELECTION_INTERVAL_MS_CONFIG, value = "1"),
+            @ClusterConfigProperty(key = GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "2")
+        }
+    )
+    public void testRebootstrap(ClusterInstance cluster) throws Exception {
+        boolean useRebootstrapTriggerMs = true;
+        int part = 0;
+        Properties producerProps = new Properties();
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        producerProps.putAll(clientConfig(useRebootstrapTriggerMs));
+
+
+        try (Admin admin = cluster.admin();
+            Producer<String, String> producer = cluster.producer(Utils.propsToMap(producerProps))) {
+
+            Properties topicProps = new Properties();
+            topicProps.put(TopicConfig.UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG, "true");
+
+            admin.createTopics(List.of(new NewTopic(topicName, 2, (short) 2).configs(Utils.propsToStringMap(topicProps)))).all().get();
```

Comment:
Should we need `cluster.waitForTopic(topicName, 2);`  here?

Reply:
I think that's only necessary when the topic metadata has changed?

Reply:
Yes, but create topic will change metadata, right?

Reply:
Oh thanks, you're right. 
But seems the old test infra doesn't wait for the metadata propagation explicitly.
So I think it's ok the skip that.
https://github.com/apache/kafka/blob/772aa241b2f1c3e8c8b7ad73ddc910c8272f2d1d/core/src/test/scala/unit/kafka/utils/TestUtils.scala#L380-L406